File: hub/src/light/effect/color_effects/linear_color_transition.py
from light.effect.partial_effect import PartialEffect
from light.effect.resolved_effect import ResolvedEffect
from util.color_transition import ColorTransition
from light.effect.effect_priority import EffectPriority

class LinearColorTransition(PartialEffect):
    def __init__(self, srcRgb, dstRgb, duration):
        super().__init__(0)
        self.src = srcRgb
        self.dst = dstRgb
        self.duration = duration
    
    def getEffect(self, t):
        dt = t - self.startTime
        tNorm = dt / self.duration
        return ResolvedEffect.createRgbw(ColorTransition.linear(tNorm, self.src, self.dst))

    def isTemporal(self):
        return False

    def isComplete(self, t):
        return False

File: hub/drumlite-js/src/light/effect/color/LinearColorTransition.ts
import RGB from "../../RGB"
import PartialEffect, { EffectParameters, EffectParameter, defaultMillisecondRange } from "../PartialEffect"
import ResolvedEffect from "../../../effect/ResolvedEffect";
import ColorTransition from "../../../util/ColorTransition";
import MidiDrumNote from "../../../midi/MidiDrumNote";

export class LinearColorTransitionParams extends EffectParameters {
    effectName = "Linear Color Transition";
    typeName = "Color";

    constructor(src: RGB = new RGB(), dst: RGB = new RGB(), duration: number = 0) {
        super();
        this.params.src = new EffectParameter<RGB>("Start Color", src, {type: "rgb"});
        this.params.dst = new EffectParameter<RGB>("End Color", dst, {type: "rgb"});
        this.params.duration = new EffectParameter<number>("Duration", duration, {range: defaultMillisecondRange});
    }
}

export default class LinearColorTransition extends PartialEffect<LinearColorTransitionParams> {
    constructor(params: LinearColorTransitionParams) {
        super(params, 0);
    }
    
    public getEffect(t: number, note?: MidiDrumNote) {
        const startTime = this.getStartTime(note);
        const dt = t - startTime;
        const tNorm = dt / (this.params.params.duration.val);
        return [[ResolvedEffect.createRgb(ColorTransition.linear(tNorm, this.params.params.src.val, this.params.params.dst.val))]];
    }

    public isTemporal() {
        return false;
    }

    public isComplete(t: number) {
        return false;
    }
}
